### PR TITLE
Migrate Ballerina Crypto module

### PR DESCRIPTION
**Pull request by Maninda**
## Purpose
Migrate Ballerina Crypto module to an independent repository.

### CODE REVIEW

**Comment by ldclakmal** [comment]
@Maninda There were 45 unit test cases. But here it seems like only 11

File: .github/workflows/build-master.yml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+          java-version: 1.8
+      - name: Grant execute permission for gradlew
+        run: chmod +x gradlew
+      - name: Build with Gradle
+        env:
+          GITHUB_TOKEN: ${{ secrets.BALLERINA_STDLIB_TOKEN }}
+          packageUser: ${{ github.actor }}
+          packagePAT: ${{ secrets.BALLERINA_STDLIB_TOKEN }}
+        run: |
+          ./gradlew publish
```
Comment (MadhukaHarith92) [inline .github/workflows/build-master.yml:27]:
```suggestion
          ./gradlew publish
          
```

File: .github/workflows/daily-build.yml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Build with Gradle
+        env:
+          GITHUB_TOKEN: ${{ secrets.BALLERINA_STDLIB_TOKEN }}
+          packageUser: ${{ github.actor }}
+          packagePAT: ${{ secrets.BALLERINA_STDLIB_TOKEN }}
+        run: ./gradlew.bat build
```
Comment (MadhukaHarith92) [inline .github/workflows/daily-build.yml:40]:
```suggestion
        run: ./gradlew.bat build
        
```

File: .github/workflows/publish-release.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        env:
+          GITHUB_TOKEN: ${{ secrets.BALLERINA_STDLIB_TOKEN }}
+          BALLERINA_CENTRAL_ACCESS_TOKEN: ${{ secrets.BALLERINA_CENTRAL_ACCESS_TOKEN }}
+          packageUser: ${{ github.actor }}
+          packagePAT: ${{ secrets.BALLERINA_STDLIB_TOKEN }}
+        run: |
+          NEW_VERSION=$(echo "${GITHUB_REF}" | cut -d "/" -f3)
+          echo "New version: ${NEW_VERSION}"
+          echo "Github username: ${GITHUB_ACTOR}"
+          ./gradlew -PpublishToCentral=true -Pversion=${NEW_VERSION} publish
```
Comment (MadhukaHarith92) [inline .github/workflows/publish-release.yml:29]:
```suggestion
          ./gradlew -PpublishToCentral=true -Pversion=${NEW_VERSION} publish
          
```

File: .github/workflows/pull-request.yml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Build with Gradle
+        env:
+          GITHUB_TOKEN: ${{ secrets.BALLERINA_STDLIB_TOKEN }}
+          packageUser: ${{ github.actor }}
+          packagePAT: ${{ secrets.BALLERINA_STDLIB_TOKEN }}
+        run: ./gradlew.bat build
```
Comment (MadhukaHarith92) [inline .github/workflows/pull-request.yml:38]:
```suggestion
        run: ./gradlew.bat build
        
```

File: build-config/checkstyle/build.gradle [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    enabled = false
+}
+
+artifacts.add('default', file("$project.buildDir/checkstyle.xml")) {
+    builtBy('downloadMultipleFiles')
+}
+
+artifacts.add('default', file("$project.buildDir/suppressions.xml")) {
+    builtBy('downloadMultipleFiles')
+}
```
Comment (MadhukaHarith92) [inline build-config/checkstyle/build.gradle:46]:
```suggestion
}

```

File: build.gradle [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        csv.enabled = true
+        xml.destination = new File("${buildDir}/reports/jacoco/report.xml")
+        html.destination = new File("${buildDir}/reports/jacoco/report.html")
+        csv.destination = new File("${buildDir}/reports/jacoco/report.csv")
+    }
+
+    onlyIf = {
+        true
+    }
+}
```
Comment (MadhukaHarith92) [inline build.gradle:102]:
```suggestion
}

```

File: crypto-ballerina/Ballerina.toml [hunk 7]
```diff
@@ -0,0 +1,20 @@
+[project]
+org-name = "ballerina"
+version = "0.7.0"
```
Comment (MadhukaHarith92) [inline crypto-ballerina/Ballerina.toml:3]:
```suggestion
version = "0.7.0"
version = "@toml.version@"
```

File: crypto-ballerina/Ballerina.toml [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+[platform]
+target = "java8"
+
+    [[platform.libraries]]
+    artifactId = "crypto"
+    version = "0.7.0-SNAPSHOT"
+    path = "../crypto-native/build/libs/crypto-native-0.7.0-SNAPSHOT.jar"
```
Comment (MadhukaHarith92) [inline crypto-ballerina/Ballerina.toml:11]:
```suggestion
    path = "../crypto-native/build/libs/crypto-native-0.7.0-SNAPSHOT.jar"
    path = "../crypto-native/build/libs/crypto-native-@project.version@.jar"
```

File: crypto-ballerina/Ballerina.toml [hunk 9]
```diff
@@ -0,0 +1,20 @@
+[project]
+org-name = "ballerina"
+version = "0.7.0"
+
+[platform]
+target = "java8"
+
+    [[platform.libraries]]
+    artifactId = "crypto"
+    version = "0.7.0-SNAPSHOT"
```
Comment (MadhukaHarith92) [inline crypto-ballerina/Ballerina.toml:10]:
```suggestion
    version = "0.7.0-SNAPSHOT"
    version = "@project.version@"
```

File: crypto-ballerina/Ballerina.toml [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    path = "../crypto-native/build/libs/crypto-native-0.7.0-SNAPSHOT.jar"
+    groupId = "ballerina"
+    modules = ["crypto"]
+
+    [[platform.libraries]]
+    artifactId = "commons-codec"
+    version = "1.9"
+    path = "./lib/commons-codec-1.9.jar"
+    groupId = "commons-codec"
+    modules = ["crypto"]
```
Comment (MadhukaHarith92) [inline crypto-ballerina/Ballerina.toml:20]:
```suggestion
    modules = ["crypto"]
    
```

**Review by MadhukaHarith92 — CHANGES_REQUESTED** [review]

**Review by ldclakmal — CHANGES_REQUESTED** [review]
@Maninda There were 45 unit-test cases in ballerina-lang repo. But here it seems like only 11 is migrated. Any reason to skip those?

**Comment by Maninda** [comment]
> @Maninda There were 45 unit-test cases in ballerina-lang repo. But here it seems like only 11 is migrated. Any reason to skip those?

New 11 tests are not the 1-to-1 migrations of the earlier Java tests. Content that are covered in the Java tests are covered with the 11 Testarina tests which are organized in a different way. And I have added error messages for all the assertions in new tests, as well.

**Comment by ldclakmal** [comment]
> > @Maninda There were 45 unit-test cases in ballerina-lang repo. But here it seems like only 11 is migrated. Any reason to skip those?
> 
> New 11 tests are not the 1-to-1 migrations of the earlier Java tests. Content that are covered in the Java tests are covered with the 11 Testarina tests which are organized in a different way. And I have added error messages for all the assertions in new tests, as well.

Any reason to organize it in a different way? IMO, unit test should test the behaviour of a singe function (API). With the new unit tests implemented, it tests multiple functions (APIs) in a singe unit test. IMO, it will become hard to debug if a test case get failed now.

File: settings.gradle [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ */
+
+plugins {
+    id "com.gradle.enterprise" version "3.2"
+}
+
+rootProject.name = 'crypto'
+include(':build-config:checkstyle')
+include ':crypto-native', ':crypto-ballerina'
+include 'crypto-native'
```
Comment (chamil321) [inline settings.gradle:17]:
```suggestion
```

**Comment by Maninda** [comment]
> > > @Maninda There were 45 unit-test cases in ballerina-lang repo. But here it seems like only 11 is migrated. Any reason to skip those?
> > 
> > 
> > New 11 tests are not the 1-to-1 migrations of the earlier Java tests. Content that are covered in the Java tests are covered with the 11 Testarina tests which are organized in a different way. And I have added error messages for all the assertions in new tests, as well.
> 
> Any reason to organize it in a different way? IMO, unit test should test the behaviour of a singe function (API). With the new unit tests implemented, it tests multiple functions (APIs) in a singe unit test. IMO, it will become hard to debug if a test case get failed now.

Separated the tests into distinct test cases and now there are 51 test cases.

File: crypto-ballerina/build.gradle [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+ */
+
+import org.apache.tools.ant.taskdefs.condition.Os
+
+description = 'Ballerina - Ballerina Crypto'
+
+configurations {
+    jbalTools
+    cryptoBallerinaBirDir
+    cryptoBallerinaJar
```
Comment (Bhashinee) [inline crypto-ballerina/build.gradle:25]:
```suggestion
```

File: crypto-native/build.gradle [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+plugins {
+    id 'java'
+}
+
+description = 'Ballerina - Crypto Java Utils'
+
+dependencies {
+    compile group: 'org.ballerinalang', name: 'ballerina-time', version: "${ballerinaLangVersion}"
+    compile group: 'org.ballerinalang', name: 'ballerina-lang', version: "${ballerinaLangVersion}"
+    compile group: 'org.ballerinalang', name: 'ballerina-runtime-api', version: "${ballerinaLangVersion}"
```
Comment (Bhashinee) [inline crypto-native/build.gradle:27]:
```suggestion
    compile group: 'org.ballerinalang', name: 'ballerina-time', version: "${ballerinaLangVersion}"
    compile group: 'org.ballerinalang', name: 'ballerina-lang', version: "${ballerinaLangVersion}"
    compile group: 'org.ballerinalang', name: 'ballerina-runtime-api', version: "${ballerinaLangVersion}"
```
Do we need these dependancies? If not can we remove them?

File: crypto-native/build.gradle [hunk 14]
```diff
@@ -0,0 +1,33 @@
...
+    compile group: 'org.ballerinalang', name: 'ballerina-java', version: "${ballerinaLangVersion}"
+    compile group: 'commons-codec', name: 'commons-codec', version: "${commoncodecVersion}"
+    compile group: 'org.ballerinalang', name: 'ballerina-runtime', version: "${ballerinaLangVersion}"
+    compile group: 'org.ballerinalang', name: 'ballerina-system', version: "${ballerinaLangVersion}"
+    compile group: 'org.slf4j', name: 'slf4j-jdk14', version: "${slf4jVersion}"
```
Comment (Bhashinee) [inline crypto-native/build.gradle:32]:
```suggestion
```
Do we need these dependencies? If not can we remove them?

File: crypto-native/build.gradle (hunk 13 above)
Comment (Bhashinee) [inline crypto-native/build.gradle:27]: Do we need these dependencies?

File: crypto-ballerina/src/crypto/tests/unit-tests/hash-function-test.bal [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/test;
+
+@test:Config {}
+function testHashMd5() {
+    byte[] input = "Ballerina test".toBytes();
+    string expectedMd5Hash = "3B12196DB784CD9F86CC635D32764FDF".toLowerAscii();
+    test:assertEquals(crc32b(input), "d37b9692", msg = "Error while Hash with CRC32b.");
```
Comment (ldclakmal) [inline crypto-ballerina/src/crypto/tests/unit-tests/hash-function-test.bal:23]:
```suggestion
    test:assertEquals(hashMd5(input), expectedMd5Hash, msg = "Error while Hash with MD5.");
```

File: crypto-ballerina/src/crypto/tests/unit-tests/hash-function-test.bal [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+}
+
```
Comment (ldclakmal) [inline crypto-ballerina/src/crypto/tests/unit-tests/hash-function-test.bal:25]: The test case for `crc32b1 API is missing here. Shall we add that also?

File: crypto-ballerina/src/crypto/tests/unit-tests/aes-cbc-encrypt-decrypt-test.bal [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+    i = 0;
+    while(i < 16) {
+        iv[i] = <byte> i;
+        i = i + 1;
+    }
+    byte[] cipherText = checkpanic encryptAesCbc(message, key, iv, "PKCS5");
+    byte[] plainText = checkpanic decryptAesCbc(cipherText, key, iv, "PKCS5");
+    test:assertEquals(plainText.toBase16(), message.toBase16(),
+        msg = "Error while Encrypt/Decrypt with AES CBC PKCS5.");
+}
```
Comment (ldclakmal) [inline crypto-ballerina/src/crypto/tests/unit-tests/aes-cbc-encrypt-decrypt-test.bal:132]: The negative test case for testing encrypt and decrypt with AES CBC PKCS1 is missing here. Shall we add that also?

File: crypto-ballerina/src/crypto/tests/unit-tests/aes-gcm-encrypt-decrypt.bal [hunk 18]
```diff
@@ -0,0 +1,108 @@
...
+    }
+    byte[]|error result = encryptAesGcm(message, invalidKey, iv, NONE, 128);
+    if (result is error) {
+        test:assertEquals(extractErrorMessage(result), "Invalid key size. valid key sizes in bytes: [16, 24, 32]",
+            msg = "Incorrect error for invalid key while No Padding Encryption with AES GCM.");
+    } else {
+        test:assertFail(msg = "No error for invalid key while No Padding Encryption with AES GCM.");
+    }
+}
+
```
Comment (ldclakmal) [inline crypto-ballerina/src/crypto/tests/unit-tests/aes-gcm-encrypt-decrypt.bal:63]: The test case for encrypt and decrypt with AES GCM NoPadding using invalid input length is missing here. Shall we add that also?

File: crypto-ballerina/src/crypto/tests/unit-tests/aes-gcm-encrypt-decrypt.bal [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+    while(i < 16) {
+        iv[i] = <byte> i;
+        i = i + 1;
+    }
+    byte[] cipherText = checkpanic encryptAesGcm(message, key, iv, "PKCS5");
+    byte[] plainText = checkpanic decryptAesGcm(cipherText, key, iv, "PKCS5");
+    test:assertEquals(plainText.toBase16(), message.toBase16(),
+        msg = "Error while Encrypt/Decrypt with AES GCM PKCS5.");
+}
+
```
Comment (ldclakmal) [inline crypto-ballerina/src/crypto/tests/unit-tests/aes-gcm-encrypt-decrypt.bal:84]: The negative test case for encrypt and decrypt with AES GCM PKCS1 is missing here. Shall we add that also?

File: crypto-ballerina/src/crypto/tests/unit-tests/rsa-ecb-encrypt-decrypt-test.bal [hunk 20]
```diff
@@ -0,0 +1,143 @@
...
+    PrivateKey invalidPrk = {algorithm:"RSA"};
+    byte[]|error result = encryptRsaEcb(message, invalidPrk, "PKCS1");
+    if (result is error) {
+        test:assertEquals(extractErrorMessage(result), "Uninitialized private/public key",
+            msg = "Incorrect error for for invalid key while Encryption with RSA ECB.");
+    } else {
+        test:assertFail(msg = "No error for invalid key Encryption with RSA ECB PKCS1.");
+    }
+}
+
```
Comment (ldclakmal) [inline crypto-ballerina/src/crypto/tests/unit-tests/rsa-ecb-encrypt-decrypt-test.bal:136]: The test case for encrypt and decrypt with RSA ECB PKCS1 with invalid padding is missing here. Shall we add that also?

File: crypto-ballerina/src/crypto/tests/unit-tests/hmac-test.bal [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+
+@test:Config {}
+function testHmacSha512() {
+    byte[] message = "Ballerina HMAC test".toBytes();
+    byte[] key = "abcdefghijk".toBytes();
+    string expectedSha512Hash = ("78d99bf3e5277fc893af6cd6b0487c33ed3abc4f956fdd1fada302f135b012a" +
+        "3c71cadaaeb462e51ff281202bdfa8807719b91f69742c3f71f036c469ac5b918").toLowerAscii();
+    test:assertEquals(hmacSha512(message, key).toBase16(), expectedSha512Hash, msg = "Error while HMAC with SHA512.");
+    test:assertTrue((trap hmacSha512(message, [])) is error, msg = "No error for empty key while HMAC with SHA512.");
+}
```
Comment (ldclakmal) [inline crypto-ballerina/src/crypto/tests/unit-tests/hmac-test.bal:65]: Shall we add a negative test case for one of the algorithms of HMAC? It was there as a test case to test HMAC generation with empty password using SHA1 algorithm.

File: crypto-ballerina/src/crypto/tests/unit-tests/aes-ecb-encrypt-decrypt-test.bal [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+    while(i < 16) {
+        key[i] = <byte> i;
+        i = i + 1;
+    }
+    message = "Ballerina crypto test".toBytes();
+    byte[] cipherText = checkpanic encryptAesEcb(message, key, "PKCS5");
+    byte[] plainText = checkpanic decryptAesEcb(cipherText, key, "PKCS5");
+    test:assertEquals(plainText.toBase16(), message.toBase16(),
+        msg = "Error while Encrypt/Decrypt with AES ECB PKCS5.");
+}
```
Comment (ldclakmal) [inline crypto-ballerina/src/crypto/tests/unit-tests/aes-ecb-encrypt-decrypt-test.bal:84]: The negative test case for testing encrypt and decrypt with AES ECV PKCS1 is missing here. Shall we add that also?

File: crypto-ballerina/src/crypto/tests/unit-tests/hash-function-test.bal (hunk 15 above)
Comment (Maninda) [reply crypto-ballerina/src/crypto/tests/unit-tests/hash-function-test.bal]: Committed.

File: crypto-ballerina/src/crypto/tests/unit-tests/hash-function-test.bal (hunk 16 above)
Comment (Maninda) [reply crypto-ballerina/src/crypto/tests/unit-tests/hash-function-test.bal:32]: Corrected.

File: crypto-ballerina/src/crypto/tests/unit-tests/aes-gcm-encrypt-decrypt.bal (hunk 18 above)
Comment (Maninda) [reply crypto-ballerina/src/crypto/tests/unit-tests/aes-gcm-encrypt-decrypt.bal:87]: As discussed offline added the disabled test till further analysis.

File: crypto-ballerina/src/crypto/tests/unit-tests/hmac-test.bal (hunk 21 above)
Comment (Maninda) [reply crypto-ballerina/src/crypto/tests/unit-tests/hmac-test.bal:92]: Extracted the negative test assertions as new test cases.

File: crypto-native/build.gradle (hunk 13 above)
Comment (Maninda) [reply crypto-native/build.gradle]: Removed.

**Review by ldclakmal — APPROVED** [review]
LGTM
